File: app/src/main/java/aleksandrkim/ArchComponentsTest/Db/Note.java
package aleksandrkim.ArchComponentsTest.Db;

import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.PrimaryKey;
import android.support.v7.util.DiffUtil;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

/**
 * Created by Aleksandr Kim on 11 Apr, 2018 6:27 PM for ArchComponentsTest
 */

@Entity
public class Note {

    @PrimaryKey(autoGenerate = true)
    private int id;
    private String title;
    private String content;
    private int color;
    private long lastModified;
    private long createdTime;

    @Ignore
    private SimpleDateFormat simpleDateFormat = new SimpleDateFormat("dd MMM", new Locale("Ru"));

    public Note(String title, String content, int color) {
        this.title = title;
        this.content = content;
        this.color = color;
        this.lastModified = Calendar.getInstance().getTimeInMillis();
        this.createdTime = Calendar.getInstance().getTimeInMillis();
    }

    public void setTitleIfEmpty () {
        if (getTitle().trim().isEmpty()) {
            String content = getContent().concat(" ");
            setTitle(content.substring(0, content.indexOf(' ')));
        }
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public int getColor() {
        return color;
    }

    public void setColor(int color) {
        this.color = color;
    }

    public String getLastModifiedString() {
        return simpleDateFormat.format(lastModified);
    }

    public String getCreatedTimeString() {
        return simpleDateFormat.format(createdTime);
    }

    public long getLastModified() {
        return lastModified;
    }

    public void setLastModified(long lastModified) {
        this.lastModified = lastModified;
    }

    public long getCreatedTime() {
        return createdTime;
    }

    public void setCreatedTime(long createdTime) {
        this.createdTime = createdTime;
    }

    public static final DiffUtil.ItemCallback<Note> DIFF_ITEM_CALLBACK = new DiffUtil.ItemCallback<Note>() {
        @Override
        public boolean areItemsTheSame(Note oldItem, Note newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(Note oldItem, Note newItem) {
            return oldItem.equals(newItem);
        }
    };

    private boolean equals(Note obj) {
        return this.id == obj.id && this.title.equals(obj.getTitle()) && this.content.equals(obj.getContent())
                && this.color == obj.getColor() && this.lastModified == obj.getLastModified();
    }
}

File: app/src/main/java/aleksandrkim/ArchComponentsTest/Db/Note.kt
package aleksandrkim.ArchComponentsTest.Db

import android.arch.persistence.room.Entity
import android.arch.persistence.room.Ignore
import android.arch.persistence.room.Index
import android.arch.persistence.room.PrimaryKey
import android.support.v7.util.DiffUtil
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by Aleksandr Kim on 11 Apr, 2018 6:27 PM for ArchComponentsTest
 */

@Entity(indices = [
    Index("id"),
    Index(value = ["title", "content"], name = "search_text")])
data class Note(var title: String = "", var content: String = "", var color: Int = -1) {

    @PrimaryKey(autoGenerate = true)
    var id: Int = 0
    var lastModified: Long = Calendar.getInstance().timeInMillis
    var createdTime: Long = Calendar.getInstance().timeInMillis

    @Ignore
    private val simpleDateFormat = SimpleDateFormat("dd MMM", Locale("Ru"))

    val lastModifiedString: String
        get() = simpleDateFormat.format(lastModified)

    val createdTimeString: String
        get() = simpleDateFormat.format(createdTime)

    private fun equals(obj: Note): Boolean = (this.id == obj.id && this.title == obj.title && this.content == obj.content
            && this.color == obj.color && this.lastModified == obj.lastModified)

    companion object {

        val DIFF_ITEM_CALLBACK: DiffUtil.ItemCallback<Note> = object : DiffUtil.ItemCallback<Note>() {
            override fun areItemsTheSame(oldItem: Note, newItem: Note): Boolean = oldItem.id == newItem.id

            override fun areContentsTheSame(oldItem: Note, newItem: Note): Boolean = oldItem.equals(newItem)
        }
    }
}
